File: ChompMan.Tests/CollisionTests.vb
Imports Microsoft.VisualStudio.TestTools.UnitTesting
Imports ChompMan.GameEngine

Namespace ChompMan.Tests

    ''' <summary>
    ''' Tests for collision detection: player vs pellets, player vs ghosts
    ''' (including frightened mode ghost eating).
    ''' </summary>
    <TestClass>
    Public Class CollisionTests

        Private Shared Function BuildOpenState() As GameState
            Dim layout = "#####" & vbLf &
                         "#...#" & vbLf &
                         "#.P.#" & vbLf &
                         "#...#" & vbLf &
                         "#####"
            Dim maze As New Maze(layout)
            Dim player As New Player(maze.PlayerStart)
            Dim ghosts As New List(Of Ghost)()
            Return New GameState(maze, player, ghosts)
        End Function

        ' ── Pellet collection ────────────────────────────────────────────────

        <TestMethod>
        Public Sub WalkingOverPellet_ConsumesIt()
            Dim state = BuildOpenState()
            Dim pelletPos As New Position(2, 3)
            state.Maze.SetCell(pelletPos, CellType.Pellet)

            Dim engine As New Engine(state)
            engine.SetInput(Direction.Right)
            For i = 1 To state.Player.SpeedTicks
                engine.Tick()
            Next

            Assert.AreEqual(CellType.Empty, state.Maze.GetCell(pelletPos))
        End Sub

        <TestMethod>
        Public Sub WalkingOverPellet_IncrementsScore()
            Dim state = BuildOpenState()
            Dim pelletPos As New Position(2, 3)
            state.Maze.SetCell(pelletPos, CellType.Pellet)

            Dim engine As New Engine(state)
            engine.SetInput(Direction.Right)
            For i = 1 To state.Player.SpeedTicks
                engine.Tick()
            Next

            Assert.AreEqual(Engine.PelletScore, state.Player.Score)
        End Sub

        <TestMethod>
        Public Sub WalkingOverPowerPellet_IncrementsScoreByPowerAmount()
            Dim state = BuildOpenState()
            Dim ppPos As New Position(2, 3)
            state.Maze.SetCell(ppPos, CellType.PowerPellet)

            Dim engine As New Engine(state)
            engine.SetInput(Direction.Right)
            For i = 1 To state.Player.SpeedTicks
                engine.Tick()
            Next

            Assert.AreEqual(Engine.PowerPelletScore, state.Player.Score)
        End Sub

        <TestMethod>
        Public Sub WalkingOverPowerPellet_FrightensGhosts()
            Dim state = BuildOpenState()
            Dim ppPos As New Position(2, 3)
            state.Maze.SetCell(ppPos, CellType.PowerPellet)

            Dim ghost As New Ghost(0, "Blinky", New Position(1, 1))
            Dim ghosts As New List(Of Ghost) From {ghost}
            Dim maze As New Maze("#####" & vbLf &
                                 "#...#" & vbLf &
                                 "#.P.#" & vbLf &
                                 "#...#" & vbLf &
                                 "#####")
            Dim player As New Player(maze.PlayerStart)
            Dim state2 As New GameState(maze, player, ghosts)
            state2.Maze.SetCell(ppPos, CellType.PowerPellet)

            Dim engine As New Engine(state2)
            engine.SetInput(Direction.Right)
            For i = 1 To player.SpeedTicks
                engine.Tick()
            Next

            Assert.AreEqual(GhostMode.Frightened, state2.Ghosts(0).Mode)
        End Sub

        ' ── Player–ghost collision ────────────────────────────────────────────

        <TestMethod>
        Public Sub TouchingNormalGhost_LosesLife()
            Dim maze As New Maze("#####" & vbLf &
                                 "#...#" & vbLf &
                                 "#.P.#" & vbLf &
                                 "#...#" & vbLf &
                                 "#####")
            Dim player As New Player(maze.PlayerStart)
            Dim ghost As New Ghost(0, "Blinky", player.Position)
            ghost.Mode = GhostMode.Chase
            Dim ghosts As New List(Of Ghost) From {ghost}
            Dim state As New GameState(maze, player, ghosts)
            Dim startLives = player.Lives
            Dim engine As New Engine(state)

            engine.Tick()

            Assert.AreEqual(startLives - 1, state.Player.Lives)
        End Sub

        <TestMethod>
        Public Sub TouchingNormalGhost_TriggersRespawn()
            Dim maze As New Maze("#####" & vbLf &
                                 "#...#" & vbLf &
                                 "#.P.#" & vbLf &
                                 "#...#" & vbLf &
                                 "#####")
            Dim player As New Player(maze.PlayerStart)
            Dim ghost As New Ghost(0, "Blinky", player.Position)
            ghost.Mode = GhostMode.Chase
            Dim ghosts As New List(Of Ghost) From {ghost}
            Dim state As New GameState(maze, player, ghosts)
            Dim engine As New Engine(state)

            engine.Tick()

            Assert.AreEqual(GamePhase.Respawning, state.Phase)
        End Sub

        <TestMethod>
        Public Sub TouchingFrightenedGhost_EatsGhost_NotLoseLife()
            Dim maze As New Maze("#####" & vbLf &
                                 "#...#" & vbLf &
                                 "#.P.#" & vbLf &
                                 "#...#" & vbLf &
                                 "#####")
            Dim player As New Player(maze.PlayerStart)
            Dim ghost As New Ghost(0, "Blinky", player.Position)
            ghost.Frighten(200)
            Dim ghosts As New List(Of Ghost) From {ghost}
            Dim startLives = player.Lives
            Dim state As New GameState(maze, player, ghosts)
            Dim engine As New Engine(state)

            engine.Tick()

            Assert.AreEqual(startLives, state.Player.Lives)
            Assert.AreEqual(GhostMode.Eaten, state.Ghosts(0).Mode)
        End Sub

        <TestMethod>
        Public Sub TouchingFrightenedGhost_AddsScore()
            Dim maze As New Maze("#####" & vbLf &
                                 "#...#" & vbLf &
                                 "#.P.#" & vbLf &
                                 "#...#" & vbLf &
                                 "#####")
            Dim player As New Player(maze.PlayerStart)
            Dim ghost As New Ghost(0, "Blinky", player.Position)
            ghost.Frighten(200)
            Dim ghosts As New List(Of Ghost) From {ghost}
            Dim state As New GameState(maze, player, ghosts)
            Dim engine As New Engine(state)

            engine.Tick()

            Assert.IsTrue(state.Player.Score >= Engine.GhostBaseScore)
        End Sub

        ' ── Level complete ────────────────────────────────────────────────────

        <TestMethod>
        Public Sub CollectingLastPellet_TriggersLevelComplete()
            Dim layout = "#####" & vbLf &
                         "#   #" & vbLf &
                         "# P #" & vbLf &
                         "#   #" & vbLf &
                         "#####"
            Dim maze As New Maze(layout)
            Dim pelletPos As New Position(2, 3)
            maze.SetCell(pelletPos, CellType.Pellet)

            Dim player As New Player(maze.PlayerStart)
            Dim ghosts As New List(Of Ghost)()
            Dim state As New GameState(maze, player, ghosts)
            Dim engine As New Engine(state)

            engine.SetInput(Direction.Right)
            For i = 1 To player.SpeedTicks
                engine.Tick()
            Next

            Assert.AreEqual(GamePhase.LevelComplete, state.Phase)
        End Sub

    End Class

End Namespace

File: ChompMan.Tests/CollisionTests.cs
using ChompMan.GameEngine;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace ChompMan.Tests;

[TestClass]
public class CollisionTests
{
    private static GameState BuildOpenState()
    {
        var layout = "#####\n#...#\n#.P.#\n#...#\n#####";
        var maze = new Maze(layout);
        var player = new Player(maze.PlayerStart);
        var ghosts = new List<Ghost>();
        return new GameState(maze, player, ghosts);
    }

    [TestMethod]
    public void WalkingOverPellet_ConsumesIt()
    {
        var state = BuildOpenState();
        var pelletPos = new Position(2, 3);
        state.Maze.SetCell(pelletPos, CellType.Pellet);
        var engine = new Engine(state);
        engine.SetInput(Direction.Right);
        for (var i = 1; i <= state.Player.SpeedTicks; i++) engine.Tick();
        Assert.AreEqual(CellType.Empty, state.Maze.GetCell(pelletPos));
    }

    [TestMethod]
    public void WalkingOverPellet_IncrementsScore()
    {
        var state = BuildOpenState();
        var pelletPos = new Position(2, 3);
        state.Maze.SetCell(pelletPos, CellType.Pellet);
        var engine = new Engine(state);
        engine.SetInput(Direction.Right);
        for (var i = 1; i <= state.Player.SpeedTicks; i++) engine.Tick();
        Assert.AreEqual(Engine.PelletScore, state.Player.Score);
    }

    [TestMethod]
    public void WalkingOverPowerPellet_IncrementsScoreByPowerAmount()
    {
        var state = BuildOpenState();
        var ppPos = new Position(2, 3);
        state.Maze.SetCell(ppPos, CellType.PowerPellet);
        var engine = new Engine(state);
        engine.SetInput(Direction.Right);
        for (var i = 1; i <= state.Player.SpeedTicks; i++) engine.Tick();
        Assert.AreEqual(Engine.PowerPelletScore, state.Player.Score);
    }

    [TestMethod]
    public void WalkingOverPowerPellet_FrightensGhosts()
    {
        var ppPos = new Position(2, 3);
        var ghost = new Ghost(0, "Blinky", new Position(1, 1));
        var ghosts = new List<Ghost> { ghost };
        var maze = new Maze("#####\n#...#\n#.P.#\n#...#\n#####");
        var player = new Player(maze.PlayerStart);
        var state = new GameState(maze, player, ghosts);
        state.Maze.SetCell(ppPos, CellType.PowerPellet);
        var engine = new Engine(state);
        engine.SetInput(Direction.Right);
        for (var i = 1; i <= player.SpeedTicks; i++) engine.Tick();
        Assert.AreEqual(GhostMode.Frightened, state.Ghosts[0].Mode);
    }

    [TestMethod]
    public void TouchingNormalGhost_LosesLife()
    {
        var maze = new Maze("#####\n#...#\n#.P.#\n#...#\n#####");
        var player = new Player(maze.PlayerStart);
        var ghost = new Ghost(0, "Blinky", player.Position) { Mode = GhostMode.Chase };
        var ghosts = new List<Ghost> { ghost };
        var state = new GameState(maze, player, ghosts);
        var startLives = player.Lives;
        var engine = new Engine(state);
        engine.Tick();
        Assert.AreEqual(startLives - 1, state.Player.Lives);
    }

    [TestMethod]
    public void TouchingNormalGhost_TriggersRespawn()
    {
        var maze = new Maze("#####\n#...#\n#.P.#\n#...#\n#####");
        var player = new Player(maze.PlayerStart);
        var ghost = new Ghost(0, "Blinky", player.Position) { Mode = GhostMode.Chase };
        var ghosts = new List<Ghost> { ghost };
        var state = new GameState(maze, player, ghosts);
        var engine = new Engine(state);
        engine.Tick();
        Assert.AreEqual(GamePhase.Respawning, state.Phase);
    }

    [TestMethod]
    public void TouchingFrightenedGhost_EatsGhost_NotLoseLife()
    {
        var maze = new Maze("#####\n#...#\n#.P.#\n#...#\n#####");
        var player = new Player(maze.PlayerStart);
        var ghost = new Ghost(0, "Blinky", player.Position);
        ghost.Frighten(200);
        var ghosts = new List<Ghost> { ghost };
        var startLives = player.Lives;
        var state = new GameState(maze, player, ghosts);
        var engine = new Engine(state);
        engine.Tick();
        Assert.AreEqual(startLives, state.Player.Lives);
        Assert.AreEqual(GhostMode.Eaten, state.Ghosts[0].Mode);
    }

    [TestMethod]
    public void TouchingFrightenedGhost_AddsScore()
    {
        var maze = new Maze("#####\n#...#\n#.P.#\n#...#\n#####");
        var player = new Player(maze.PlayerStart);
        var ghost = new Ghost(0, "Blinky", player.Position);
        ghost.Frighten(200);
        var ghosts = new List<Ghost> { ghost };
        var state = new GameState(maze, player, ghosts);
        var engine = new Engine(state);
        engine.Tick();
        Assert.IsTrue(state.Player.Score >= Engine.GhostBaseScore);
    }

    [TestMethod]
    public void CollectingLastPellet_TriggersLevelComplete()
    {
        var layout = "#####\n#   #\n# P #\n#   #\n#####";
        var maze = new Maze(layout);
        var pelletPos = new Position(2, 3);
        maze.SetCell(pelletPos, CellType.Pellet);
        var player = new Player(maze.PlayerStart);
        var ghosts = new List<Ghost>();
        var state = new GameState(maze, player, ghosts);
        var engine = new Engine(state);
        engine.SetInput(Direction.Right);
        for (var i = 1; i <= player.SpeedTicks; i++) engine.Tick();
        Assert.AreEqual(GamePhase.LevelComplete, state.Phase);
    }
}
